File: WD.HabitTracker.Infra.Persistence.Mongodb/Documents/Users/UserDocument.cs
using MongoDB.Bson.Serialization.Attributes;
using TWD.HabitTracker.Domain.Entities.User;
using WD.HabitTracker.Infra.Persistence.Mongodb.Documents.Users.Auth;

namespace WD.HabitTracker.Infra.Persistence.Mongodb.Documents.Users;

public class UserDocument
{
    public UserDocument(User user)
    {
        Id = user.Id;
        AuthInfo = new AuthInfoDocument(user.AuthInfo);
    }

    public int SchemaVersion { get; set; } = 1;
    
    [BsonId]
    public Guid Id { get; set; }
    
    [BsonElement]
    public AuthInfoDocument AuthInfo { get; set; }

    public User ToUser() => new (Id, AuthInfo.ToAuthInfo());
}

File: WD.HabitTracker.Persistence.MongoDb/Documents/User.Document.fs
namespace WD.HabitTracker.Persistence.MongoDb.Documents

open System
open MongoDB.Bson.Serialization.Attributes

[<BsonIgnoreExtraElements>]
[<CLIMutable>]
type CredentialsAuthDocument = {
    Email: string
    Password: string
}

[<BsonIgnoreExtraElements>]
[<CLIMutable>]
type  DeviceAuthDocument = {
    DeviceToken: string
}

[<BsonIgnoreExtraElements>]
[<CLIMutable>]
type AuthInfoDocument = {
    [<BsonElement>] [<BsonIgnoreIfNull>] DeviceAuth: DeviceAuthDocument
    [<BsonElement>] [<BsonIgnoreIfNull>] CredentialsAuth: CredentialsAuthDocument
}

[<BsonIgnoreExtraElements>]
[<CLIMutable>]
type UserDocument = {
    SchemaVersion: int
    [<BsonId>] [<BsonIgnoreIfDefault>] Id: Guid
    [<BsonElement>] [<BsonIgnoreIfNull>] AuthInfo: AuthInfoDocument
}

